Python/chapter03/p03_stack_of_plates/miguelHx.py

```diff
@@ -0,0 +1,374 @@
+    """Stack data structure implementation.
+    Uses LIFO (last-in first-out) ordering.
+    The most recent item added to the stack is
+    the first removed.  Traversal is top to bottom.
+    """
+
+    def __init__(self):
+        self.top: Optional[StackNode[T]] = None # top is a pointer to StackNode object
+        self.bottom: Optional[StackNode[T]] = None
+        self.size: int = 0
+        self.current_node: Optional[StackNode[T]] = self.top
+        return
+
+    def pop(self) -> T:
+        """
+        Removes the top item from the stack
+        Raises:
+            IndexError: raised when pop is attempted on empty stack
+        Returns:
+            int: The data at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty.')
+        item = self.top.data
+        self.top = self.top.next
+        if self.size > 1 and self.top:
+            self.top.prev = None
+        self.size -= 1
+        return item
+
+    def push(self, item: T) -> None:
+        """
+        Adds an item to the top of the stack
+        Args:
+            item (int): data we want at the top of stack
+        """
+        t = StackNode(item, None, None)
+        t.next = self.top
+        self.top = t
+        if self.size == 0:
+            self.bottom = t
+        elif t.next:
+            t.next.prev = t
+        self.size += 1
+
+    def peek(self) -> T:
+        """
+        Returns data at the top of the stack
+        Raises:
+            IndexError: [description]
+        Returns:
+            int: the value at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty')
+        return self.top.data
+
+    def __iter__(self) -> Iterator:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        return MyStackIterator(self.top, self.bottom, self.size)
+
+    def __bool__(self) -> bool:
+        """
+        True is returned when the container is not empty.
+        From https://docs.python.org/3/reference/datamodel.html#object.__bool__ :
+        Called to implement truth value testing and the built-in operation bool();
+        should return False or True. When this method is not defined, len() is called,
+        if it is defined, and the object is considered true if its result is nonzero.
+        If a class defines neither len() nor bool(), all its instances are considered true.
+        Returns:
+            bool: False when empty, True otherwise
+        """
+        return self.size > 0
+    
+    def __len__(self) -> int:
+        return self.size
+
+    def __str__(self):
+        if self.size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
```

Comment:
`mypy` is complaining that `n` could be `None`:

```
$ mypy  miguelHx.py
miguelHx.py:148: error: Item "None" of "Optional[StackNode[T]]" has no attribute "next"
miguelHx.py:149: error: Item "None" of "Optional[StackNode[T]]" has no attribute "data"
miguelHx.py:150: error: Item "None" of "Optional[StackNode[T]]" has no attribute "next"
miguelHx.py:151: error: Item "None" of "Optional[StackNode[T]]" has no attribute "data"
Found 4 errors in 1 file (checked 1 source file)
```

Reply:
d9d28ccce9e7770e42dcf942375414e2a92bc471